File: src/ghsingo/internal/audio/bellbank.go
package audio

import (
	"fmt"
	"os"
	"path/filepath"
)

// BellBank maps each pentatonic Pitch to either a pre-rendered long-form sample
// (preferred for rank-1 and Release events) or an on-demand Karplus-Strong
// synth voice (used for rank 2-4 and as a fallback when a sample is missing).
type BellBank struct {
	sampleRate int
	samples    map[Pitch][]float32
	synthDecay float64
}

// NewBellBank creates an empty bank using the given sample rate for synth.
func NewBellBank(sampleRate int) *BellBank {
	return &BellBank{
		sampleRate: sampleRate,
		samples:    make(map[Pitch][]float32),
		synthDecay: 0.996,
	}
}

// SetSynthDecay configures the Karplus-Strong decay coefficient (0.990~0.999).
func (b *BellBank) SetSynthDecay(decay float64) {
	b.synthDecay = decay
}

// LoadFromDir scans dir for {note}{octave}.wav files matching the 15 pitches
// and loads every one it finds. Missing files are silently skipped (the
// runtime falls back to synth). Returns the number of samples loaded.
func (b *BellBank) LoadFromDir(dir string) (int, error) {
	loaded := 0
	for _, p := range AllPitches() {
		path := filepath.Join(dir, p.Filename())
		if _, err := os.Stat(path); err != nil {
			continue
		}
		pcm, err := LoadWavFile(path)
		if err != nil {
			return loaded, fmt.Errorf("load %s: %w", path, err)
		}
		b.samples[p] = pcm
		loaded++
	}
	return loaded, nil
}

// HasSample reports whether a pre-rendered sample exists for pitch p.
func (b *BellBank) HasSample(p Pitch) bool {
	_, ok := b.samples[p]
	return ok
}

// SampleVoice returns a new playback voice for p, or nil if no sample is
// loaded for that pitch.
func (b *BellBank) SampleVoice(p Pitch, velocity float32) *SampleVoice {
	pcm, ok := b.samples[p]
	if !ok {
		return nil
	}
	return &SampleVoice{pcm: pcm, velocity: velocity}
}

// SynthVoice returns a new Karplus-Strong synth voice for p.
func (b *BellBank) SynthVoice(p Pitch, velocity float32) *KarplusVoice {
	return NewKarplusVoice(b.sampleRate, p.Frequency(), b.synthDecay, velocity)
}

// SampleVoice is a one-shot sampled-bell playback instance.
type SampleVoice struct {
	pcm      []float32
	pos      int
	velocity float32
}

// NextSample returns one sample and advances the playhead.
func (s *SampleVoice) NextSample() float32 {
	if s.pos >= len(s.pcm) {
		return 0
	}
	v := s.pcm[s.pos] * s.velocity
	s.pos++
	return v
}

// Done reports whether the sample has finished.
func (s *SampleVoice) Done() bool {
	return s.pos >= len(s.pcm)
}

File: src/rsghsing/src/audio/bellbank.rs
//! Accent voice bank. Ported from `internal/audio/bellbank.go`.
//!
//! Call-chain note for the P2 report: `render-audio-v2` loads the sampled
//! bank via `LoadFromDir`, but `MixerV2.spawnAccent` only ever calls
//! `SynthVoice` — `SampleVoice` is never constructed on the render path. The
//! loaded PCM is therefore inert; it is kept so the port's behaviour (and the
//! "load N samples" log line) matches Go exactly.

use std::path::Path;

use anyhow::Result;

use crate::audio::karplus::KarplusVoice;
use crate::audio::pitch::{self, Pitch};
use crate::audio::wav;

pub struct BellBank {
    sample_rate: i32,
    samples: Vec<(Pitch, Vec<f32>)>,
    synth_decay: f64,
}

impl BellBank {
    pub fn new(sample_rate: i32) -> Self {
        BellBank {
            sample_rate,
            samples: Vec::new(),
            synth_decay: 0.996,
        }
    }

    pub fn set_synth_decay(&mut self, decay: f64) {
        self.synth_decay = decay;
    }

    /// Scans `dir` for `{note}{octave}.wav` matching the 15 pitches.
    /// Missing files are skipped. Returns the number loaded.
    pub fn load_from_dir(&mut self, dir: &Path) -> Result<usize> {
        let mut loaded = 0;
        for p in pitch::all_pitches() {
            let path = dir.join(p.filename());
            if !path.exists() {
                continue;
            }
            let pcm = wav::load_wav_file(&path)
                .map_err(|e| anyhow::anyhow!("load {}: {e}", path.display()))?;
            self.samples.push((p, pcm));
            loaded += 1;
        }
        Ok(loaded)
    }

    pub fn has_sample(&self, p: Pitch) -> bool {
        self.samples.iter().any(|(k, _)| *k == p)
    }

    pub fn synth_voice(&self, p: Pitch, velocity: f32, seed: i64) -> KarplusVoice {
        KarplusVoice::new(self.sample_rate, p.frequency(), self.synth_decay, velocity, seed)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn synth_voice_is_silenced_after_two_seconds() {
        let b = BellBank::new(44_100);
        let mut v = b.synth_voice(
            Pitch { note: pitch::Note::Gong, octave: pitch::Octave::Mid },
            1.0,
            1,
        );
        for _ in 0..44_100 * 3 {
            v.next_sample();
        }
        assert!(v.done());
        assert_eq!(v.next_sample(), 0.0);
    }

    #[test]
    fn load_from_dir_missing_dir_loads_nothing() {
        let mut b = BellBank::new(44_100);
        assert_eq!(b.load_from_dir(Path::new("/nonexistent-rsghsing")).unwrap(), 0);
    }
}
